### PR TITLE
Fix for Issue #26

### PR DESCRIPTION
**Pull request by MaxSac**
Return Secondaries as shared Pointer. Maybe Secondaries should get some utility functions.
Secondaries are pybinded. To get the vector of dynamicData, "secondaries.particles" has to
be called.

### CODE REVIEW

**Comment by MaxSac** [comment]
Memory fix cause another unattractive problem. The garbage collector for each `secondaries` is only called when the program is terminated. That makes it slow and consumes a lot of memory.
```
import pyPROPOSAL as pp

prop = pp.Propagator( particle_def=pp.particle.MuMinusDef.get(), config_file="PATH_TO_CONFIG/config.json")
mu = prop.particle
mu_init = pp.particle.Particle(mu)
mu_init.energy = 1e9
mu_backup.direction = pp.Vector3d(0,0,-1)

for i in range(1000):
    mu.inject_state(mu_backup)
    secondaries = prop.propagate(1e6)
    print(secodaries.particles[0].id)
```
@MaxNoe is there a way to avoid the problem, than explicitly calling the garbage can collector?

**Comment by MaxNoe** [comment]
For loops in python don't have scope. Functions do. So if you use a function to propagate a single particle, the secondaries should be cleared up. Or just using `del secondaries`.

You can check memory consumption using psutils;

```
process = psutil.Process(os.getpid())
print(f'Currently using {p.memory_info().rss / 1024**2:.1f} MB RAM')
```

**Comment by MaxNoe** [comment]
One question I have: why is this a shared pointer on a new custom class that basically just wraps a vector? why not `std::vector<shared_ptr<Particle>>` or what ever would be appropriate?

That would have the advantage, that you do not have to reimplement all the methods and protocols of vector. It iteration won't work currently.

**Comment by MaxNoe** [comment]
Sorry, wrong button

File: public/PROPOSAL/Secondaries.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#pragma once
+
+
+#include "PROPOSAL/particle/Particle.h"
+#include <memory>
+#include <vector>
+
+namespace PROPOSAL {
+
+class Secondaries {
```
Comment (MaxNoe) [inline public/PROPOSAL/Secondaries.h]: Why this wrapper class? Why not just use the `std::vector<std::shared_ptr<DynamicData>>` directly?

**Comment by MaxSac** [comment]
> One question I have: why is this a shared pointer on a new custom class that basically just wraps a vector? why not `std::vector<shared_ptr<Particle>>` or what ever would be appropriate?
> 
> That would have the advantage, that you do not have to reimplement all the methods and protocols of vector. It iteration won't work currently.

There should be some querys on the wrapper class which are not yet implemented.
The pointer itself is needless and i will delete it in next commit.

File: public/PROPOSAL/particle/ParticleDef.h [hunk 2]
```diff
@@ -332,29 +332,29 @@ PROPOSAL_MAKE_HASHABLE(PROPOSAL::ParticleDef, t.mass, t.lifetime, t.charge)
 namespace PROPOSAL {
 static std::map<const int, const ParticleDef&> Id_Particle_Map
```
Comment (MaxNoe) [inline public/PROPOSAL/particle/ParticleDef.h]: wouln't `const ParticleType` as type for the map be better than all those casts to int?

**Pull request merged**